File: 46-permutations/permutations.java
class Solution {

    public void solve(int idx, int nums[],List<Integer> curr, List<List<Integer>> ans, boolean vis[] )
    {
        if(curr.size()==nums.length)
        {
            ans.add(new ArrayList<>(curr));
            return;
        }


        for(int i=idx; i<nums.length; i++)
        {
            if(vis[i]==true)
            continue;

            curr.add(nums[i]);
            vis[i]=true;
            solve(idx,nums,curr,ans,vis);
            curr.remove(curr.size()-1);
            vis[i]=false;

        }
    }
    public List<List<Integer>> permute(int[] nums) {
        List<Integer> curr= new ArrayList<>();
        List<List<Integer>> ans= new ArrayList<>();

        boolean vis[]= new boolean[nums.length];

        solve(0,nums,curr,ans,vis);
        return ans;
    }
}

File: 46-permutations/permutations.cpp
class Solution {
private:
    void helper(vector<vector<int>> &ans,vector<int> &nums,vector<int> &ds,vector<bool> &vis){
       
        if(ds.size() == nums.size()){
            ans.push_back(ds);
            return ;
        }

        for(int i = 0 ; i < nums.size() ; i++){
            if(!vis[i]){
                ds.push_back(nums[i]);
                vis[i] = 1;
                helper(ans,nums,ds,vis);
                ds.pop_back();
                vis[i] = 0;
            }
        }

    }
public:
    vector<vector<int>> permute(vector<int>& nums) {
        int n = nums.size();
        vector<int> ds;
        vector<vector<int>> ans;
        vector<bool> vis(n);
        helper(ans,nums,ds,vis);
        return ans;
    }
};